spacy/kb_base.pyx

```diff
@@ -0,0 +1,168 @@
+# cython: infer_types=True, profile=True
+
+# todo
+#   - submit PR
+#   - add tests (w/ dummy class)
+#   - update docs
+from pathlib import Path
+from typing import Iterable, Tuple, Iterator, Union
+from cymem.cymem cimport Pool
+
+from .tokens import Span
+from .util import SimpleFrozenList
+
+
+cdef class Candidate:
+    """A `Candidate` object refers to a textual mention (`alias`) that may or may not be resolved
+    to a specific `entity` from a Knowledge Base. This will be used as input for the entity linking
+    algorithm which will disambiguate the various candidates to the correct one.
+    Each candidate (alias, entity) pair is assigned to a certain prior probability.
+
+    DOCS: https://spacy.io/api/kb/#candidate_init
+    """
+
+    def __init__(self, BaseKnowledgeBase kb, entity_hash, entity_freq, entity_vector, alias_hash, prior_prob):
+        self.kb = kb
+        self.entity_hash = entity_hash
+        self.entity_freq = entity_freq
+        self.entity_vector = entity_vector
+        self.alias_hash = alias_hash
+        self.prior_prob = prior_prob
+
+    @property
+    def entity(self):
+        """RETURNS (uint64): hash of the entity's KB ID/name"""
+        return self.entity_hash
+
+    @property
+    def entity_(self):
+        """RETURNS (str): ID/name of this entity in the KB"""
+        return self.kb.vocab.strings[self.entity_hash]
+
+    @property
+    def alias(self):
+        """RETURNS (uint64): hash of the alias"""
+        return self.alias_hash
+
+    @property
+    def alias_(self):
+        """RETURNS (str): ID of the original alias"""
+        return self.kb.vocab.strings[self.alias_hash]
+
+    @property
+    def entity_freq(self):
+        return self.entity_freq
+
+    @property
+    def entity_vector(self):
+        return self.entity_vector
+
+    @property
+    def prior_prob(self):
+        return self.prior_prob
+
+
+def get_candidates_batch(kb: BaseKnowledgeBase, mentions: Iterable[Union[Span, str]]) -> Iterable[Iterator[Candidate]]:
+    """
+    Return candidate entities for the given mentions and fetching appropriate entries from the index.
+    kb (BaseKnowledgeBase): Knowledge base to query.
+    mention (Iterable[Union[Span, str]]): Entity mentions for which to identify candidates.
+    RETURNS (Iterable[Iterator[Candidate]]): Identified candidates.
+    """
+    return kb.get_candidates_batch(mentions)
```

Comment:
I went back and forth about whether it makes sense to add a swappable method like `get_candidates()` for batches. In the end I went for it, mostly for consistency. I don't like the current situation that much because we have an mostly useless layer of abstraction here. Ideally we want users to work with `BaseKnowledgeBase.get_candidates()` and `BaseKnowledgeBase.get_candidates_batch()` directly. Actually just the batched version, but getting rid of the non-batched version would obviously break stuff.

So we are redirecting calls once more than necessary or useful. This might be confusing for users, but this seems like the least bad/confusing, most consistent alternative to me that is also non-breaking. Open to other ideas. 

In v4 I'd toss out the non-batched version and the `get_candidates`/`get_candidates_batch` config options.

Reply:
So basically the conflict we've got here is being able to overwrite some of the functionality in the config, versus providing inheritance as a means to overwrite that same functionality.

I guess our initial reasoning was to allow users to only overwrite the candidate generation without touching the KB class itself. But as we've seen, the two are often so tightly coupled that maybe that didn't make so much sense, or was covering only a subpart of the problem efficiently.

At this point (v3) as you've said we can't remove the config option, so adding a batched one feels like the most consistent to me as well. I also assume that most users won't actually see the internal code structure here so hopefully it shouldn't be too confusing.

Reply:
I think this is similar with what we went through adding patterns with the Matcher, and this does look like the best solution. In v4 I think it'd be good to go back to the simple name (`get_candidates`) but make it clear the input type should be a list.

Reply:
Agree to returning to `get_candidates` in v4.